crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -481,50 +487,81 @@ impl InstancePool {
+        unsafe {
+            std::ptr::drop_in_place(instance as *mut _);
+        }
+        // The instance is now uninitialized memory and cannot be
+        // touched again until we write a fresh Instance in-place with
+        // std::ptr::write in allocate() above.
 
         self.free_list.lock().unwrap().push(index);
     }
 
     fn set_instance_memories(
+        instance_idx: usize,
         instance: &mut Instance,
-        mut memories: impl Iterator<Item = *mut u8>,
+        memories: &MemoryPool,
+        maybe_memfds: &Option<Arc<ModuleMemFds>>,
         max_pages: u64,
     ) -> Result<(), InstantiationError> {
         let module = instance.module.as_ref();
 
         debug_assert!(instance.memories.is_empty());
 
-        for plan in
-            (&module.memory_plans.values().as_slice()[module.num_imported_memories..]).iter()
+        for (memory_index, plan) in module
+            .memory_plans
+            .iter()
+            .skip(module.num_imported_memories)
         {
+            let defined_index = module
+                .defined_memory_index(memory_index)
+                .expect("should be a defined memory since we skipped imported ones");
+
             let memory = unsafe {
                 std::slice::from_raw_parts_mut(
-                    memories.next().unwrap(),
+                    memories.get_base(instance_idx, memory_index),
                     (max_pages as usize) * (WASM_PAGE_SIZE as usize),
                 )
             };
-            instance.memories.push(
-                Memory::new_static(plan, memory, commit_memory_pages, unsafe {
-                    &mut *instance.store()
-                })
-                .map_err(InstantiationError::Resource)?,
-            );
+
+            if let Some(memfds) = maybe_memfds {
+                let image = memfds.get_memory_image(defined_index);
+                let mut slot = memories.take_memfd_slot(instance_idx, memory_index)?;
+                let initial_size = plan.memory.minimum * WASM_PAGE_SIZE as u64;
+
+                // If instantiation fails, the whole slot in the
+                // pooling allocator's mmap region is permanently in a
+                // bad state, because we could not alter its mapping
+                // for some reason. We can't continue like this -- in
+                // the best case, the system is in a degraded state
+                // where one slot is permanently out of
+                // commission. It's better to panic the whole process,
+                // because we have no way of returning an error that
+                // means "please throw away the whole allocator" (and
+                // dropping its mapping might fail again later too,
+                // anyway).
+                slot.instantiate(initial_size as usize, image).expect(
+                    "Fatal error in setting up slot;
+                             pool mmap is in unknown condition, cannot continue",
+                );
```

Comment:
I feel like our error handling here is sort of half-measures in the sense that if this initialization fails then an abort happens but if `Memory::new_static` fails below it's bubbled up and handled. In an error below (or later for a memory afterwards) it seems like it would still result in dropping the `MemFdSlot` which itself doesn't do any cleanup and leaves the slot in a somewhat indeterminate state?

In some sense this may not be the end of the world really about having an indeterminate state because allocation of a memfd slot here always redoes the entire mapping for the entire memory, so if something is "leaked" it's only temporary in the sense that no wasm will be using that slot?

Reply:
That's a good point. On thinking about this further:

- If instantiation bails out after we've initialized the MemFdSlot (which can happen, e.g., if the resource-limiter says "no") then it's most idiomatic to just drop the slot object and let the pool create a new one later.

- The most important thing -- the only thing I can think of that would absolutely require us to panic the process -- is that we don't let some other mmap ASLR its way into a gap in the pooling allocator's region if we leave a hole somewhere during instantiation. But we already avoid that, to prevent exactly that race condition; so even if the MemFdSlot fails, say, the second of 3 mmaps, we are still covering/reserving the whole range.

- Given that, I think it's fine to drop the MemFdSlot at any point.

- For an extra layer of {sanity, avoiding resource leaks e.g. keeping a memfd image alive, defense-in-depth-don't-map-more-than-we-need}, it is still nice to try to remove a stale mapping. I've written a `Drop` handler for `MemFdSlot` that does this in a best-effort way: it tries to map anonymous memory with `PROT_NONE` over the whole slot. It uses `NORESERVE` to make this more likely to succeed. But if it doesn't, the error is explicitly ignored (with a comment in the source to describe why safe): nothing is *supposed* to touch the slot anyway until the next `MemFdSlot` is created for it, and if mapping something into the slot fails again later for the new `MemFdSlot`, then it has the proper avenues to bubble up the error at that time.

Does that seem reasonable?